Mathlib/CategoryTheory/ChosenFiniteProducts.lean

```diff
@@ -195,6 +195,263 @@ instance (priority := 100) : Limits.HasFiniteProducts C :=
+`prodComparison`. -/
+@[simps]
+def prodComparisonNatTrans (A : C) :
+    (curriedTensor C).obj A ⋙ F ⟶ F ⋙ (curriedTensor D).obj (F.obj A) where
+  app B := prodComparison F A B
+  naturality x y f := by
+    apply hom_ext <;>
+    simp only [Functor.comp_obj, curriedTensor_obj_obj,
+      Functor.comp_map, curriedTensor_obj_map, Category.assoc, prodComparison_fst, whiskerLeft_fst,
+      prodComparison_snd, prodComparison_snd_assoc, whiskerLeft_snd, ← F.map_comp]
+
+/-- The product comparison morphism from `F(- ⊗ -)` to `F- ⊗ F-`, whose components are given by
+`prodComparison`. -/
+@[simps]
+def prodComparisonBifunctorNatTrans :
+    curriedTensor C ⋙ (whiskeringRight _ _ _).obj F ⟶
+      F ⋙ curriedTensor D ⋙ (whiskeringLeft _ _ _).obj F where
+  app A := prodComparisonNatTrans F A
+  naturality x y f := by
+    ext z
+    apply hom_ext <;>
+    simp only [Functor.comp_obj, whiskeringRight_obj_obj, curriedTensor_obj_obj,
+      whiskeringLeft_obj_obj, Functor.comp_map, whiskeringRight_obj_map, NatTrans.comp_app,
+      whiskerRight_app, curriedTensor_map_app, prodComparisonNatTrans_app, Category.assoc,
+      prodComparison_fst, whiskeringLeft_obj_map, whiskerLeft_app, whiskerRight_fst,
+      prodComparison_fst_assoc, prodComparison_snd, whiskerRight_snd, ← F.map_comp]
+
+
+instance (A : C) [∀ B, IsIso (prodComparison F A B)] : IsIso (prodComparisonNatTrans F A) := by
+  letI : ∀ X, IsIso ((prodComparisonNatTrans F A).app X) := by assumption
+  apply NatIso.isIso_of_isIso_app
+
+instance [h : ∀ A B, IsIso (prodComparison F A B)] : IsIso (prodComparisonBifunctorNatTrans F) := by
+  letI : ∀ X, IsIso ((prodComparisonBifunctorNatTrans F).app X) :=
+    fun _ ↦ (by dsimp; apply NatIso.isIso_of_isIso_app)
+  apply NatIso.isIso_of_isIso_app
+
+/-- The natural isomorphism `F(A ⊗ -) ≅ FA ⊗ F-`, provided each `prodComparison F A B` is an
+isomorphism (as `B` changes). -/
+@[simps!]
+noncomputable def prodComparisonNatIso (A : C)
+    [∀ B, IsIso (prodComparison F A B)] :
+    (curriedTensor C).obj A ⋙ F ≅ F ⋙ (curriedTensor D).obj (F.obj A) :=
+  asIso (prodComparisonNatTrans F A)
+
+@[simp]
+lemma prodComparisonNatIso_hom (A : C) [∀ B, IsIso (prodComparison F A B)] :
+    (prodComparisonNatIso F A).hom = prodComparisonNatTrans F A := rfl
+
+/-- The natural isomorphism of bifunctors `F(- ⊗ -) ≅ F- ⊗ F-`, provided each
+`prodComparison F A B` is an isomorphism. -/
+@[simps!]
+noncomputable def prodComparisonBifunctorsNatIso
+    [∀ A B, IsIso (prodComparison F A B)] :
+    curriedTensor C ⋙ (whiskeringRight _ _ _).obj F ≅
+      F ⋙ curriedTensor D ⋙ (whiskeringLeft _ _ _).obj F :=
+  asIso (prodComparisonBifunctorNatTrans F)
+
+@[simp]
+lemma prodComparisonBifunctorsNatIso_hom [∀ A B, IsIso (prodComparison F A B)] :
+    (prodComparisonBifunctorsNatIso F).hom = prodComparisonBifunctorNatTrans F := rfl
+
+theorem prodComparison_comp {E : Type u₂} [Category.{v₂} E] [ChosenFiniteProducts E] (G : D ⥤ E) :
```

Comment:
You may also add a `prodComparison_id` lemma for the identity functor.

Reply:
Yes. For completeness I also added `_comp` and `_id` lemmas for the natural transforms as well.